core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -203,6 +204,52 @@ private[spark] class SparkSubmit extends Logging {
     }
   }
 
+  /**
+   * Tries to read the mainClass from the JAR manifest if not already set.
+   *
+   * Works with non-local JARs as well.
+   *
+   * @param args Spark submit arguments
+   * @param hadoopConf Hadoop configuration
+   * @param jarPath Path to JAR file, can be remote
+   * @return the FQDN of the main class or null if not found
+   */
+  private def resolveMainClassIfNeeded(
+    args: SparkSubmitArguments,
+    hadoopConf: HadoopConfiguration,
+    jarPath: String
+  ): String = {
+    if (args.mainClass != null) {
+      return args.mainClass
+    }
+
+    var mainClass: String = null
+    if (args.mainClass == null && !args.isPython && !args.isR && jarPath != null) {
```

Comment:
Checking `args.mainClass` is redundant here. It also feels like the other checks for R and Python don't really belong here.

This method would be better if all it did was get the main class from a jar file (regardless of whether the file is local or remote).